python/private/repl.bzl

```diff
@@ -33,26 +20,54 @@ _generate_repl_main = rule(
     attrs = {
         "out": attr.output(
             mandatory = True,
+            doc = "The path to the file to generate.",
         ),
         "stub": attr.label(
             mandatory = True,
             allow_single_file = True,
+            doc = ("The stub responsible for actually invoking the final shell. " +
+                   "See the \"Customizing the REPL\" docs for details."),
         ),
         "_template": attr.label(
             default = "//python/private:repl_template.py",
             allow_single_file = True,
+            doc = "The template to use for generating `out`.",
         ),
     },
+    doc = """\
+Generates a "main" script for a py_binary target that starts a Python REPL.
+
+The template is designed to take care of the majority of the logic. The user
+customizes the exact shell that will be started via the stub. The stub is a
+simple shell script that imports the desired shell and then executes it.
+""",
 )
 
 def py_repl_binary(name, stub, deps = [], data = [], **kwargs):
+    """A py_binary target that executes a REPL when run.
+
+    The stub is the script that ultimately decides which shell the REPL will run.
+    It can be as simple as this:
+
+        import code
+        code.interact()
+
+    Or it can load something like IPython instead.
+
+    Args:
+        name: Name of the generated py_binary target.
+        stub: The script that invokes the shell.
+        deps: The dependencies of the py_binary.
+        data: The runtime dependencies of the py_binary.
+        **kwargs: Forwarded to the py_binary.
+    """
     _generate_repl_main(
         name = "%s_py" % name,
         stub = stub,
         out = "%s.py" % name,
     )
 
-    py_binary(
+    _py_binary(
```

Comment:
nit: I remember @rickeylev mentioning to avoid `out` parameters because then we can use providers to pass the files around more cleverly. I think the `_generate_repl_main` could do that, but if you disagree on this approach, then it is fine to leave as is here for now.

Reply:
Done

It did make the filename a bit weird though. Instead of `repl.py`, it's now called `repl_py.py`. Will have to ask Richard about that. Unless I misunderstood what you had in mind, apologies.